### PR TITLE
Add deploy -t/--travis to deploy via Travis

### LINKED ISSUE #457: Provide a way for github repos to auto-deploy docs using Travis
Current: the user has to specifically run the `markbind deploy` to deploy the site to GitHub. This is suitable for user docs which should only be deployed when a new version of the product has been released

Suggestion: Also provide a way to set up Travis so that docs are deployed everytime a specific branch is updated (e.g., `master`). 

Justification: When using Markbind for dev docs, docs needs to be auto-updated every time the master branch is updated.

Side note: in the longer term, it's ideal if ther is a way to maintain both user docs and dev docs in the same repo but update user docs only when there is a new release yet update dev docs everytime the master branch updates. Note sure if we can do that though. Until we get there, we can say that user docs site should be deployed to a separate repo.

### PR DESCRIPTION
**Pull request by Xenonym**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Enhancement to an existing feature

Resolves #457.

**What is the rationale for this request?**
Currently, MarkBind supports deploying to GitHub Pages via `markbind deploy`. This is suitable for user documentation that only needs to be published with every new version. However, developer documentation is usually updated more frequently, and it will be useful to support a `-t/--travis` option to allow automatic deployment upon a push via Travis CI.

**What changes did you make? (Give an overview)**
I added a `-t/--travis [tokenVar]` option that can be run in Travis CI to automatically build and deploy a MarkBind site. `-t/--travis [tokenVar]` will:

1. Check that MarkBind is currently being run in Travis CI.
2. Retrieve the GitHub access token stored in `GITHUB_TOKEN` or a user specified `tokenVar` enviornment variable.
3. Extract the GitHub repo slug if the user has specified a repo in `site.json`, and construct a HTTPS URL containing the token for pushing changes.
4. Build and deploy the site.

**Is there anything you'd like reviewers to focus on?**
1. Currently, I assume that we are only deploying to GitHub. This implementation may need to change for other Git providers, specifically, the regex for extracting the repo slug and the authentication method (not sure if other Git providers do `https://[token]@provider/repo`) .

**Testing instructions:**
1. Create a MarkBind site on GitHub and add it to Travis CI. Sample instructions [here](https://deploy-preview-649--markbind-master.netlify.com/userguide/deployingthesite#deploying-to-git-hub-pages-via-travis-ci).
2. Add an appropriate `.travis.yml` to run `markbind deploy -t`. A sample `.travis.yml`:
(Differs from the one in the documentation as it does not restrict to only building from `master`.)
```yaml
language: node_js
node_js:
  - '8'
install:
  - npm i -g git://github.com/Xenonym/markbind.git#deploy-travis
script: markbind deploy --travis
```
3. Push some changes to trigger the Travis CI build. The site should deploy to GitHub Pages automatically.

### CODE REVIEW

File: docs/userGuide/deployingTheSite.md [hunk 1]
```diff
@@ -49,6 +49,40 @@ You can override the default deployment settings %%(e.g., repo/branch to deploy)
...
+**You can setup [<tooltip content="a platform for Continous Integration and Delivery (among other things)">Travis CI</tooltip>](https://www.travis-ci.org/) to automatically build and deploy your site on GitHub Pages every time your GitHub repo is updated.**
+
+Here are the steps to set up Travis CI:
+
+1. [Sign up with GitHub](https://travis-ci.com/signin) at [Travis CI](https://travis-ci.org).
+1. Accept the autorisation for Travis CI when you are redirected to GitHub.
+1. Click the green _Activate_ button, and select the repository with the MarkBind site.
+1. [Generate a GitHub personal access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/#creating-a-token) with **repo** permissions. Take note of the generated token - you will not be able to see it again once you navigate away from the page.
+1. [Add an environment variable in Travis CI](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings) named `GITHUB_TOKEN`, with the value set to the personal access token generated in the previous step. Ensure that _Display value in the build log_ is set to _Off_.
```
Comment (damithc) [inline docs/userGuide/deployingTheSite.md]: Add this screenshot (or something similar) here?
image

File: docs/userGuide/deployingTheSite.md [hunk 2]
```diff
@@ -49,6 +49,40 @@ You can override the default deployment settings %%(e.g., repo/branch to deploy)
...
+1. Add a `.travis.yml` file to instruct Travis CI to run `markbind deploy --travis` when you push to the repository. An example `.travis.yml` file that can accomplish this is given below:
+    <box background-color="white" border-color="white">
+    ```yaml
+    language: node_js
+    node_js:
+      - '8'
+    install:
+      - npm i -g markbind-cli
+    script: markbind deploy --travis
```
Comment (nicholaschuayunzhi) [inline docs/userGuide/deployingTheSite.md]: Is it possible to do a markbind build && markbind deploy --travis here? That way deploy would not have to generate. The `-t` option would simply mean deploying in a travis environment?

**Comment by nicholaschuayunzhi** [comment]
Could you write some tests too? Thank you!

File: src/Site.js [hunk 3]
```diff
@@ -813,6 +814,38 @@ Site.prototype.deploy = function () {
...
+          }
+
+          const githubToken = process.env[travisTokenVar];
+          let repoSlug = process.env.TRAVIS_REPO_SLUG;
+          if (options.repo) {
+            // Extract repo slug from user-specified repo URL so that we can include the access token
+            const repoSlugRegex = /github\.com[:/]([\w-]+\/[\w-]+)\.git/;
+            const repoSlugMatch = repoSlugRegex.exec(options.repo);
+            if (!repoSlugMatch) {
+              reject(new Error(`The specified repository ${options.repo} is not a valid GitHub repository.`));
```
Comment (nicholaschuayunzhi) [inline src/Site.js]: Our normal deploy works for different git providers. Could we make this error more specific for the travis build? Something like:

`Deploy with travis expects a GitHub repository. The specified repository ${options.repo} is invalid.`

**Comment by Xenonym** [comment]
@nicholaschuayunzhi changed the behaviour of `deploy --travis` to only deploy, not build, as well as added some tests! Also changed some of the deploy tests to have more consistent descriptions, please see if its okay!

**Comment by damithc** [comment]
Managed to set it up for https://nus-cs3281.github.io/2019/ 
Good work @Xenonym 

A side question: is it possible to set up Travis to use the latest minor version of MarkBind, e.g., `v2.*`? That way, provided we follow semantic versioning, Travis will use the latest version that will not have any breaking changes.

**Comment by Xenonym** [comment]
> A side question: is it possible to set up Travis to use the latest minor version of MarkBind, e.g., `v2.*`? That way, provided we follow semantic versioning, Travis will use the latest version that will not have any breaking changes.

In a pinch, you can install `markbind-cli@^1.18.0`, which will install the latest version `>=1.18.0` and `< 2.0.0` based on [`semver` syntax](https://github.com/npm/node-semver#caret-ranges-123-025-004):

```yaml
install:
  - npm i -g markbind-cli@^1.18.0
```

Alternatively, if we anticipate this to be a common use case, we should [tag the latest `v1.x` version on `npm`](https://docs.npmjs.com/adding-dist-tags-to-packages), eg. `v1.x-latest`. Then, a user can just do `npm install -g markbind-cli@v1.x-latest`.

**Comment by damithc** [comment]
> @damithc In a pinch, you can install `markbind-cli@^1.18.0`, which will install the latest version `>=1.18.0` and `< 2.0.0` based on [`semver` syntax](https://github.com/npm/node-semver#caret-ranges-123-025-004):

Nice. I think this is good enough for the time being.

**Pull request merged**